flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java

```diff
@@ -0,0 +1,445 @@
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobmanager.OnCompletionActions;
+import org.apache.flink.runtime.jobmaster.factories.JobMasterServiceFactory;
+import org.apache.flink.runtime.leaderelection.LeaderContender;
+import org.apache.flink.runtime.leaderelection.LeaderElectionService;
+import org.apache.flink.runtime.messages.Acknowledge;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.function.FunctionUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The runner for the job manager. It deals with job level leader election and make underlying job manager
+ * properly reacted.
+ */
+public class JobManagerRunnerImpl implements LeaderContender, OnCompletionActions, JobManagerRunner {
+
+	private static final Logger log = LoggerFactory.getLogger(JobManagerRunnerImpl.class);
+
+	// ------------------------------------------------------------------------
+
+	/** Lock to ensure that this runner can deal with leader election event and job completion notifies simultaneously. */
+	private final Object lock = new Object();
+
+	/** The job graph needs to run. */
+	private final JobGraph jobGraph;
+
+	/** Used to check whether a job needs to be run. */
+	private final RunningJobsRegistry runningJobsRegistry;
+
+	/** Leader election for this job. */
+	private final LeaderElectionService leaderElectionService;
+
+	private final LibraryCacheManager libraryCacheManager;
+
+	private final Executor executor;
+
+	private final JobMasterService jobMasterService;
+
+	private final FatalErrorHandler fatalErrorHandler;
+
+	private final CompletableFuture<ArchivedExecutionGraph> resultFuture;
+
+	private final CompletableFuture<Void> terminationFuture;
+
+	private CompletableFuture<Void> leadershipOperation;
+
+	/** flag marking the runner as shut down. */
+	private volatile boolean shutdown;
+
+	private volatile CompletableFuture<JobMasterGateway> leaderGatewayFuture;
+
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Exceptions that occur while creating the JobManager or JobManagerRunner are directly
+	 * thrown and not reported to the given {@code FatalErrorHandler}.
+	 *
+	 * @throws Exception Thrown if the runner cannot be set up, because either one of the
+	 *                   required services could not be started, or the Job could not be initialized.
+	 */
+	public JobManagerRunnerImpl(
+			final JobGraph jobGraph,
```

Comment:
It could be a possible discussion that whether we can get rid of `final` modifier of method parameters. It often(always?) does harm when we directly mutate parameters and developers always think a parameter is `final`.

If we said all parameter should be effective final then we can gain both immutability and code simplicity.

Reply:
I agree that this is not done very consistently in the Flink code base. Do you want to start a dev ML discussion about it and add the the result to Flink's coding guide?

Reply:
Yes I could start a discussion tomorrow :-)